### PR TITLE
Circles Rebuild - Part XIV - Another Contribution Amount Validation Bug-fix

### PR DESCRIPTION
**Pull request by Ap0c**
## Why are you doing this?

@CPKING has found another one.

### Problem

When the user had entered an invalid custom amount, and then switched to a contribution type where it was a valid amount, the error still showed until they clicked the custom amount field, at which point it was re-parsed.

### Solution

Rework the parsing mechanism to always store custom amounts if they are valid numbers, even if they aren't within the bounds of the given contribution type. This has the side-benefit of showing the entered amount in the payment button, even if it doesn't fit in the bounds, instead of resetting it to a default.

[**Trello Card**](https://trello.com/c/asGbLr0v/1314-bug-validation-on-us-page)

cc @JustinPinner 

## Changes

- Simplified the parsing of custom amounts by adding a new parsing function. The old one now just exists for the old contributions component, and will go away when the non-circles pages do.
- Added a test to catch the bug.

### CODE REVIEW

File: assets/helpers/contributions.js [hunk 1]
```diff
@@ -94,6 +99,22 @@ function parse(input: ?string, contrib: Contrib): ParsedContrib {
 
 }
 
+function betterParse(input: string, contributionType: Contrib): ParsedAmount {
```
Comment (svillafe) [inline assets/helpers/contributions.js]: Is there a more meaningful name than `betterParse`? What are we parsing here?

Comment (svillafe) [reply assets/helpers/contributions.js]: If it is an improvement over the parse method, can we replace that method?

Comment (Ap0c) [reply assets/helpers/contributions.js]: "The old one now just exists for the old contributions component, and will go away when the non-circles pages do."

Comment (svillafe) [reply assets/helpers/contributions.js]: Fair enough, Is it possible to come up with a  better name?

**Pull request merged**